README.md
## PyUQ

This is a python client library for [UQ](https://github.com/buaazp/uq) cluster.


CHANGELOG.md
Changelog
=========

### 0.1.1
- Add get_many mock function for memcache client to fix pop key sequence disorder bug


tests/build_uq.sh
#!/bin/sh -e

ORG_PATH="github.com/buaazp"
REPO_PATH="${ORG_PATH}/uq"

export GOPATH=${PWD}/gopath

go get -u $REPO_PATH

cd /tmp
git clone https://$REPO_PATH
cd uq
go build
cp uq /tmp/bin/


tests/test_api.py
# coding: utf-8

import unittest
import datetime

import uq


class TestUqClientApi(unittest.TestCase):

    def check_default_resp(self, resp):
        self.assertIsInstance(resp, tuple)
        self.assertEqual(len(resp), 2)
        self.assertEqual(resp[0], True)
        self.assertEqual(resp[1], '')

    def _test_api(self, protocol, port):
        cli = uq.UqClient(protocol=protocol, ip='localhost', port=port)
        r = cli.add('foo')
        self.check_default_resp(r)

        r = cli.add('foo', 'x', datetime.timedelta(seconds=10))
        self.check_default_resp(r)

        r = cli.push('foo', 'hello')
        self.check_default_resp(r)

        r = cli.pop('foo/x')
        self.assertEqual(r[2], b'hello')

        r = cli.remove(r[1])
        self.check_default_resp(r)

    def test_http_api(self):
        self._test_api(uq.ProtocolHttp, 8001)

    def test_redis_api(self):
        self._test_api(uq.ProtocolRedis, 8002)

    '''
    FIXME:
        pymemcache get_many command may rearrange the keys sequence,
        this leads uq library pop processing with error.
    '''
    # def test_memcache_api(self):
    #     self._test_api(uq.ProtocolMemcache, 8003)


tests/wait_etcd.sh
#!/bin/sh

etcd_host="127.0.0.1"
etcd_port=4001

while true; do
    port_test=$(nc -vz $etcd_host $etcd_port 2>&1|grep "Connection refused")
    if [ "$port_test" != "" ]; then
        echo "$(date) - waiting for etcd start up..."
        sleep 1
    else
        echo "Connection with etcd $etcd_host:$etcd_port successfully..."
        break
    fi
done


uq/conn.py
# coding: utf-8

import random
import logging
import functools

import etcd
import redis
import requests
from pymemcache.client.base import Client as McClient
from pymemcache.exceptions import MemcacheError, MemcacheClientError

from .exceptions import UqError, UqNotImplementedError
from .consts import MaxRetry
from .utils import timedetla_to_str


def query_addrs(etcd_cli, etcd_key):
    try:
        resp = etcd_cli.read(
            '/{}/servers'.format(etcd_key), sorted=True, recursive=False)
    except etcd.EtcdKeyNotFound as e:
        logging.error('server entrypoint not found: {}'.format(e))
        raise
    except etcd.EtcdConnectionFailed as e:
        logging.error('failed to connect etcd: {}'.format(e))
        raise

    if not resp.dir:
        raise UqError('invalid etcd key')
    for child in resp.children:
        yield child.key.split('/')[-1]


class Conn(object):
    '''
    Base class of connection from uq client to uq server.
    '''

    def __init__(self, addr, etcd_cli, etcd_key):
        self.addr = addr
        self.etcd_cli = etcd_cli
        self.etcd_key = etcd_key
        self.addrs = None
        self._update_conn_pool()

    def _choose(self):
        if len(self.addrs) == 0:
            return None
        return random.choice(self.addrs)

    def _update_conn_pool(self):
        raise UqNotImplementedError

    def _add(self, addr, data):
        raise UqNotImplementedError

    def _push(self, addr, data):
        raise UqNotImplementedError

    def _pop(self, addrs, data):
        raise UqNotImplementedError

    def _remove(self, addr, data):
        raise UqNotImplementedError

    def add(self, topic, line, recycle):
        if not topic:
            raise UqError('topic is empty')
        updated = False
        while True:
            retry = 0
            while retry < MaxRetry:
                addr = self._choose()
                data = {
                    'topic': topic,
                    'line': line,
                    'recycle': timedetla_to_str(recycle),
                }
                if addr is None:
                    logging.error('no uq server available')
                else:
                    result = self._add(addr, data)
                if result is None:
                    retry += 1
                else:
                    return result

            if not updated:
                self._update_conn_pool()
                updated = True
            else:
                break

        return False, 'all conn add failed after retry'

    def push(self, key, value):
        updated = False
        while True:
            retry = 0
            while retry < MaxRetry:
                addr = self._choose()
                if addr is None:
                    logging.error('no uq server available')
                else:
                    data = {
                        'key': key,
                        'value': value,
                    }
                    result = self._push(addr, data)
                    if result is None:
                        retry += 1
                    else:
                        return result
                retry += 1

            if not updated:
                self._update_conn_pool()
                updated = True
            else:
                break

        return False, 'all conn push failed after retry'

    def pop(self, key):
        updated = False
        while True:
            retry = 0
            while retry < MaxRetry:
                data = {'key': key}
                result = self._pop(self.addrs, data)
                if result is None:
                    retry += 1
                else:
                    return result

            if not updated:
                self._update_conn_pool()
                updated = True
            else:
                break

        return False, '', 'all conn pop failed after retry'

    def remove(self, key):
        values = key.split('/', 1)
        if len(values) != 2:
            return False, 'key illegal'
        addr, cid = values

        retry = 0
        while retry < MaxRetry:
            data = {'cid': cid}
            result = self._remove(addr, data)
            if result is None:
                retry += 1
            else:
                return result
            retry += 1

        return False, 'all conn del failed after retry'


def catch_requests_error(func):
    @functools.wraps(func)
    def wrap_f(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except requests.RequestException as e:
            logging.error('requests error: {}'.format(e))
            return None
    return wrap_f


class HttpConn(Conn):
    '''
    Http protocol based connection between uq client and uq server
    '''

    def __init__(self, addr, etcd_cli, etcd_key):
        super(HttpConn, self).__init__(addr, etcd_cli, etcd_key)

    def _update_conn_pool(self):
        if self.etcd_cli is None:
            self.addrs = [self.addr]
        else:
            self.addrs = [x for x in query_addrs(self.etcd_cli, self.etcd_key)]

    @catch_requests_error
    def _add(self, addr, data):
        url = 'http://{0}/v1/queues'.format(addr)
        r = requests.put(url, data)
        if r.status_code == requests.codes.created:
            return True, ''
        else:
            if 'Existed' in r.text:
                return False, r.text
            logging.error('add error: {}'.format(r.text))
            return None

    @catch_requests_error
    def _push(self, addr, data):
        key = data.pop('key')
        url = 'http://{0}/v1/queues/{1}'.format(addr, key)
        r = requests.post(url, data)
        if r.status_code == requests.codes.no_content:
            return True, ''
        else:
            logging.error('push error: {}'.format(r.text))
            return None

    @catch_requests_error
    def _pop(self, addrs, data):
        nomsg = 0
        for addr in addrs:
            url = 'http://{0}/v1/queues/{1}'.format(addr, data['key'])
            r = requests.get(url)
            if r.status_code == requests.codes.not_found:
                nomsg += 1
            elif r.status_code == requests.codes.ok:
                cid = '{}/{}'.format(addr, r.headers.get('X-UQ-ID'))
                value = r.text
                return True, cid, value
            else:
                logging.error('pop error: {}'.format(r.text))
        if nomsg == len(addrs):
            return False, '', 'no message'
        return None

    @catch_requests_error
    def _remove(self, addr, data):
        url = 'http://{0}/v1/queues/{1}'.format(addr, data['cid'])
        r = requests.delete(url)
        if r.status_code == requests.codes.no_content:
            return True, ''
        else:
            logging.error('del error: {}'.format(r.text))
            return None


def catch_redis_error(func):
    @functools.wraps(func)
    def wrap_f(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except redis.RedisError as e:
            logging.error('redis error: {}'.format(e))
            return None
    return wrap_f


class RedisConn(Conn):
    '''
    RESP protocol based connection between uq client and uq server
    '''

    def __init__(self, addr, etcd_cli, etcd_key):
        super(RedisConn, self).__init__(addr, etcd_cli, etcd_key)

    def _update_conn_pool(self):
        if self.etcd_cli is None:
            self.addrs = [self.addr]
        else:
            self.addrs = [x for x in query_addrs(self.etcd_cli, self.etcd_key)]
        self.conns = {}
        for addr in self.addrs:
            try:
                host, port = addr.split(':')
            except ValueError:
                raise UqError('invalid rq address {}'.format(addr))
            self.conns[addr] = redis.StrictRedis(host=host, port=port)

    @catch_redis_error
    def _add(self, addr, data):
        if not data['line']:
            args = ['ADD', data['topic']]
        else:
            full_line_name = '{topic}/{line}'.format(**data)
            args = ['ADD', full_line_name, data['recycle']]

        try:
            r = self.conns[addr].execute_command(*args)
        except redis.exceptions.ResponseError as e:
            errstr = str(e)
            if 'Existed' in errstr:
                return False, errstr
            logging.error('add error: {}'.format(errstr))
            return None
        if r == 'OK':
            return True, ''
        return None

    @catch_redis_error
    def _push(self, addr, data):
        self.conns[addr].set(data['key'], data['value'])
        return True, ''

    @catch_redis_error
    def _pop(self, addrs, data):
        nomsg = 0
        for addr in addrs:
            try:
                r = self.conns[addr].execute_command('GET', data['key'])
                cid = '{}/{}'.format(addr, r[1])
                value = r[0]
                return True, cid, value
            except redis.exceptions.ResponseError as e:
                if 'No Message' in str(e):
                    nomsg += 1
                else:
                    logging.error('pop error: {}'.format(e))
                    return None
        if nomsg == len(addrs):
            return False, '', 'no message'
        return None

    @catch_redis_error
    def _remove(self, addr, data):
        try:
            self.conns[addr].execute_command('DEL', data['cid'])
        except ValueError as e:
            if str(e) == "invalid literal for int() with base 10: 'OK'":
                # UQ returns 'OK' when messages(keys) were removed, while
                # the standard redis returns the number of successful removed
                # keys when executing DEL.
                return True, ''
            else:
                raise
        return True, ''


def catch_memcache_error(func):
    @functools.wraps(func)
    def wrap_f(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except MemcacheError as e:
            logging.error('memcache error: {}'.format(e))
            return None
    return wrap_f


class MemcacheConn(Conn):
    '''
    Memcache protocol based connection between uq client and uq server
    '''

    def __init__(self, addr, etcd_cli, etcd_key):
        super(MemcacheConn, self).__init__(addr, etcd_cli, etcd_key)

    def _update_conn_pool(self):
        if self.etcd_cli is None:
            self.addrs = [self.addr]
        else:
            self.addrs = [x for x in query_addrs(self.etcd_cli, self.etcd_key)]
        self.conns = {}
        for addr in self.addrs:
            try:
                host, port = addr.split(':')
                port = int(port)
            except ValueError:
                raise UqError('invalid rq address {}'.format(addr))
            self.conns[addr] = McClient((host, port))

    @catch_memcache_error
    def _add(self, addr, data):
        if not data['line']:
            key = data['topic']
            value = ''
        else:
            key = '{topic}/{line}'.format(**data)
            value = data['recycle']
        self.conns[addr].add(key, value)
        return True, ''

    @catch_memcache_error
    def _push(self, addr, data):
        self.conns[addr].set(data['key'], data['value'])
        return True, ''

    @catch_memcache_error
    def _pop(self, addrs, data):
        nomsg = 0
        for addr in addrs:
            try:
                key = data['key']
                r = self.conns[addr].get_many([key, 'id'])
                cid = '{}/{}'.format(addr, r['id'])
                value = r[key]
                return True, cid, value
            except MemcacheClientError as e:
                if 'No Message' in str(e):
                    nomsg += 1
                else:
                    logging.error('pop error: {}'.format(e))
                    return None
        if nomsg == len(addrs):
            return False, '', 'no message'
        return None

    @catch_memcache_error
    def _remove(self, addr, data):
        self.conns[addr].delete(data['cid'])
        return True, ''


uq/client.py
# coding: utf-8

import etcd

from .exceptions import UqError, UqNotImplementedError
from .consts import ProtocolHttp, ProtocolMemcache, ProtocolRedis
from .conn import HttpConn, RedisConn, MemcacheConn


class Client(object):
    '''
    Base class of client for the uq library.
    '''

    def __init__(self, protocol):
        if protocol not in [ProtocolHttp, ProtocolRedis, ProtocolMemcache]:
            raise UqError('invalid protocol: {}'.format(protocol))
        self.protocol = protocol
        self._init_uq_conn()

    def _init_uq_conn(self):
        raise UqNotImplementedError

    def add(self, topic, line=None, recycle=None):
        return self.conn.add(topic, line, recycle)

    def push(self, key, value):
        return self.conn.push(key, value)

    def pop(self, key):
        return self.conn.pop(key)

    def remove(self, key):
        return self.conn.remove(key)


class UqClient(Client):
    '''
    Client implementation for communication with standalone uq server
    '''

    def __init__(self, protocol='http', ip='localhost', port=8808, **kwargs):
        self._ip = ip
        self._port = port
        self._kwargs = kwargs
        super(UqClient, self).__init__(protocol)

    def _init_uq_conn(self):
        addr = '{ip}:{port}'.format(ip=self._ip, port=self._port)
        if self.protocol == ProtocolHttp:
            self.conn = HttpConn(addr, None, None)
        elif self.protocol == ProtocolRedis:
            self.conn = RedisConn(addr, None, None)
        elif self.protocol == ProtocolMemcache:
            self.conn = MemcacheConn(addr, None, None)
        else:
            raise UqError('invalid protocol: {}'.format(self.protocol))


class UqClusterClient(Client):
    '''
    Client implementation for communication with uq cluster
    '''

    def __init__(self, protocol='http', etcd_host='127.0.0.1', etcd_port=4001,
                 etcd_key='uq', **kwargs):
        self._etcd_host = etcd_host
        self._etcd_port = etcd_port
        self._etcd_key = etcd_key
        self._etcd_cli = None
        _kwargs = {}
        for k, v in kwargs.iteritems():
            # use prefix '_etcd_' to avoid name confliction with
            # positional arguments
            if k.startswith('_etcd_'):
                k = k[len('_etcd_'):]
            _kwargs[k] = v
        self._kwargs = _kwargs
        super(UqClusterClient, self).__init__(protocol)

    @property
    def etcd_cli(self):
        if self._etcd_cli is None:
            self._etcd_cli= etcd.Client(
                host=self._etcd_host, port=self._etcd_port, **self._kwargs)
        return self._etcd_cli

    def _init_uq_conn(self):
        if self.protocol == ProtocolHttp:
            self.conn = HttpConn(None, self.etcd_cli, self._etcd_key)
        elif self.protocol == ProtocolRedis:
            self.conn = RedisConn(None, self.etcd_cli, self._etcd_key)
        elif self.protocol == ProtocolMemcache:
            self.conn = MemcacheConn(None, self.etcd_cli, self._etcd_key)
        else:
            raise UqError('invalid protocol: {}'.format(self.protocol))


uq/duration.py
# coding: utf-8

'''
This module is used for converting between python datetime.timedelta and
Go's Duration strings.

copyright by https://github.com/icholy/Duration.py
'''

import re
import datetime

_nanosecond_size  = 1
_microsecond_size = 1000 * _nanosecond_size
_millisecond_size = 1000 * _microsecond_size
_second_size      = 1000 * _millisecond_size
_minute_size      = 60   * _second_size
_hour_size        = 60   * _minute_size

units = {
    "ns" : _nanosecond_size,
    "us" : _microsecond_size,
    "µs" : _microsecond_size,
    "μs" : _microsecond_size,
    "ms" : _millisecond_size,
    "s"  : _second_size,
    "m"  : _minute_size,
    "h"  : _hour_size
}


def from_str(duration):
    """Parse a duration string to a datetime.timedelta"""

    if duration in ("0", "+0", "-0"):
        return datetime.timedelta()

    pattern = re.compile('([\-\+\d\.]+)([a-zµμ]+)')
    total = 0
    sign = -1 if duration[0] == '-' else 1
    matches = pattern.findall(duration)

    if not len(matches):
        raise Exception("Invalid duration {}".format(duration))

    for (value, unit) in matches:
        if unit not in units:
            raise Exception(
                "Unknown unit {} in duration {}".format(unit, duration))
        try:
            total += float(value) * units[unit]
        except:
            raise Exception(
                "Invalid value {} in duration {}".format(value, duration))

    microseconds = total / _microsecond_size
    return datetime.timedelta(microseconds=sign * microseconds)


def to_str(delta):
    """Format a datetime.timedelta to a duration string"""

    total_seconds = delta.total_seconds()
    sign = "-" if total_seconds < 0 else ""
    nanoseconds = abs(total_seconds * _second_size)

    if total_seconds < 1:
        result_str = _to_str_small(nanoseconds)
    else:
        result_str = _to_str_large(nanoseconds)

    return "{}{}".format(sign, result_str)


def _to_str_small(nanoseconds):

    result_str = ""

    if not nanoseconds:
        return "0"

    milliseconds = int(nanoseconds / _millisecond_size)
    if milliseconds:
        nanoseconds -= _millisecond_size * milliseconds
        result_str += "{}ms".format(milliseconds)

    microseconds = int(nanoseconds / _microsecond_size)
    if microseconds:
        nanoseconds -= _microsecond_size * microseconds
        result_str += "{}us".format(microseconds)

    if nanoseconds:
        result_str += "{}ns".format(nanoseconds)

    return result_str


def _to_str_large(nanoseconds):

    result_str = ""

    hours = int(nanoseconds / _hour_size)
    if hours:
        nanoseconds -= _hour_size * hours
        result_str += "{}h".format(hours)

    minutes = int(nanoseconds / _minute_size)
    if minutes:
        nanoseconds -= _minute_size * minutes
        result_str += "{}m".format(minutes)

    seconds = float(nanoseconds) / float(_second_size)
    if seconds:
        nanoseconds -= _second_size * seconds
        result_str += "{}s".format(seconds)

    return result_str


examples/py-uq.py
# coding: utf-8

import logging
import datetime
import argparse

#  from uq.client import UqClient, UqClusterClient
import uq


def _strip(value):
    if isinstance(value, bytes):
        value = value.decode('utf-8')
    return value.strip('\r\n') if hasattr(value, 'strip') else value


def parseargs():
    parser = argparse.ArgumentParser()
    parser.add_argument('--ip', default='127.0.0.1', help='uq server ip address')
    parser.add_argument('--port', type=int, default=8808, help='uq server port')
    parser.add_argument('--protocol', default='http', choices=['redis', 'http', 'mc'], help='fronent interface')
    parser.add_argument('--etcd', default='', help='etcd service location')
    parser.add_argument('--cluster', default='uq', help='cluster name in etcd')
    parser.add_argument('--topic', default='foo', help='topic name')
    parser.add_argument('--line', default='x', help='line name')
    return parser.parse_args()


def main(args):
    if args.etcd == '':
        cli = uq.UqClient(args.protocol, args.ip, args.port)
    else:
        kwargs = {}
        if args.etcd.startswith('http'):
            kwargs['_etcd_protocol'] = args.etcd[:args.etcd.index(':')]
            netloc = args.etcd[args.etcd.index('://')+3:]
        else:
            netloc = args.etcd
        if ':' in netloc:
            etcd_host = netloc[:netloc.index(':')]
            etcd_port = int(netloc[netloc.index(':')+1:])
        else:
            etcd_host = netloc
            etcd_port = 4001
        cli = uq.UqClusterClient(
            args.protocol, etcd_host, etcd_port, args.cluster, **kwargs)

    recycle = datetime.timedelta(seconds=10)
    success, info = cli.add(args.topic, '', 0)
    if not success:
        if 'Existed' not in info:
            logging.error('add error: {}'.format(_strip(info)))
            return
    logging.info('add topic: {}'.format(_strip(info)))

    success, info = cli.add(args.topic, args.line, recycle)
    if not success:
        if 'Existed' not in info:
            logging.error('add error: {}'.format(_strip(info)))
            return
    logging.info('add topic: {}'.format(_strip(info)))

    value = 'hello'.encode()
    success, info = cli.push(args.topic, value)
    if not success:
        logging.error('push error: {}'.format(_strip(info)))
        return
    logging.info('push to topic: {}'.format(_strip(info)))

    key = args.topic + '/' + args.line
    success, cid, data = cli.pop(key)
    if not success:
        logging.error('pop error: {}'.format(_strip(data)))
        return
    logging.info('pop from topic: {} {}'.format(cid, _strip(data)))

    success, info = cli.remove(cid)
    if not success:
        logging.error('del error: {}'.format(_strip(info)))
        return
    logging.info('del from topic: {}'.format(_strip(info)))

    logging.info('all test finished')

if __name__ == '__main__':
    logging.basicConfig(
        format='%(asctime)s %(module)s/%(funcName)s %(levelname)s "%(message)s"', level=logging.INFO)
    logging.getLogger("requests").setLevel(logging.WARNING)
    logging.getLogger("urllib3").setLevel(logging.WARNING)
    args = parseargs()
    main(args)
